### PR TITLE
[ADD] l10n_jo_edi: Added Jordan EDI module

### PR DESCRIPTION
**Pull request by hsal-odoo**
This commit adds the E-invoicing app for Jordan.
The E-invoicing is mandatory for businesses in Jordan.

task-3895493

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py:121]: I see this method is normally used for the global discounts (e.g. a negative invoice line in Odoo), not the discounts per line.  (weird that it does not give errors then)

File: addons/l10n_jo_edi/models/account_move_send.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move_send.py:4]: normally in the wizard/

File: addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py:158]: I suppose the main difference with the standard UBL is that it is always in JOD

Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py:158]: Yeah I think so. They really got a standard and customized it very heavily that it's no longer a standard :"D

**Comment by derick-david** [comment]
Will this be back-ported to odoo 17 ?

**Comment by hsal-odoo** [comment]
> Will this be back-ported to odoo 17 ?

Yesss

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:108]: What happens in the case of a timeout and the invoice would have been accepted?  Do we have a way to see if an invoice was submitted? (anyways: suppose it times out on our side, but not on the gov side)

For error handling, it would be better to put everything in the "errors" key.  For the moment, the cron does not catch exceptions (it might change in the future and I see you catch it later anyways), but putting it in the errors key also gives better rendering

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:158]: Why at post and not at generation?

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:151]: The idea would be to pass what is in the message_post to error.  (passing the error object itself might be a bit harsh)

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:53]: Is it always needed in Jordan?  (no threshold?)

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [reply addons/l10n_jo_edi/models/account_move.py:158]: I mean, why at creation of the invoice (he could even still change company on the invoice) and not when generating the XML

Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/account_move.py:158]: Ah got it

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:143]: still needed?

File: addons/l10n_jo_edi/models/account_move_send.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move_send.py:8]: I don't think this method does anything more that is useful than the l10n_jo_edi_is_needed.  Maybe we keep the to_send as a dummy state, but at least do not depend on it here.  That way, at least we avoid the confusion between applicable and needed.

File: addons/l10n_jo_edi/models/account_tax.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_tax.py:10]: What if someone else wants to add an exempt tax?  We could simply add a boolean or tax type selection field.

File: addons/l10n_jo_edi/models/res_company.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/res_company.py:12]: So only when it is sales will you have refunds?

File: addons/l10n_jo_edi/tests/test_jo_edi_precision.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/tests/test_jo_edi_precision.py:47]: I would put the long docstring of the other method here.  (and refer in the method to this test)

**Review by jco-odoo — COMMENTED** [review]
I saw from the comments before about the locking in the 17 PR.  That should happen through the SELECT UPDATE FOR.  (it has nothing to do with invalidation: it is about making sure that the same invoice is not being sent at the same time by 2 different processes.)

File: addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py
Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py:133]: Yeah `_get_country_vals` is trigerred from here, and also, I don't want the extra values stored in the parent class to contaminate my XML. I want to only report expected values.

File: addons/l10n_jo_edi/models/account_move.py
Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/account_move.py:83]: The attachment file could be there, but the submit failed. In that case, the reset to draft button should be enabled. The only way the reset to draft button would be hidden is when the XML is successfully sent (l10n_jo_edi_state == 'sent')

File: addons/l10n_jo_edi/models/account_move.py
Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/account_move.py:171]: Why not? Also, the locking part is not so clear for me.

File: addons/l10n_jo_edi/models/account_move_send.py
Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/account_move_send.py:8]: Well, the state is actually needed to prevent submitting invoices more than once. If I would not rely on the state, my condition would change to be
```python
return move.l10n_jo_edi_is_needed and (move.l10n_jo_edi_error or not move.l10n_jo_edi_xml_attachment_file)
```
which is needlessly redundant

File: addons/l10n_jo_edi/models/res_company.py
Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/res_company.py:12]: No, in all taxpayer types we can have refunds.

File: addons/l10n_jo_edi/models/account_move_send.py
Comment (jco-odoo) [reply addons/l10n_jo_edi/models/account_move_send.py:8]: It is confusing

File: addons/l10n_jo_edi/models/res_company.py
Comment (jco-odoo) [reply addons/l10n_jo_edi/models/res_company.py:12]: What is the _is_sales_refund method needed for then?

File: addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py
Comment (jco-odoo) [reply addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py:199]: Please provide a docstring here: 
""" For JO UBL the document allowance charge vals needs to be the sum of the line discounts. """

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:103]: Everything was done in UTF-8 and then here you convert back in ascii?

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:106]: We can not provide the proper certificate instead of putting verify False?

Comment (jco-odoo) [reply addons/l10n_jo_edi/models/account_move.py:106]: is not params already a string?

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [reply addons/l10n_jo_edi/models/account_move.py:103]: Should not we do some kind of json dump instead as well?

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:108]:
```suggestion
            return _("Request time out! Please try again.")
```

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:70]: Are you sure there is nothing to be escaped in l10n_jo_edi_qr?

File: addons/l10n_jo_edi/tests/jo_edi_common.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/tests/jo_edi_common.py:62]: It is like you could just extend move_vals with company, partner_jo  (normally that is more the journal on invoices) put an invoice_date/currency if nothing is there, ...

Comment (hsal-odoo) [reply addons/l10n_jo_edi/tests/jo_edi_common.py:62]: That would work if I did not need to create multiple lines (with possibly multiple taxes)

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [inline addons/l10n_jo_edi/models/account_move.py:113]: @aboo-odoo Can we avoid security on this one?  @hsal-odoo Is it json, otherwise we could use the jsonrpc method.

Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/account_move.py:113]: Yeah it's json

**Review by aboo-odoo — COMMENTED** [review]
I'm having an overview of the rest of the code for security related stuff. In the meantime, I've left you a comment you can work on :+1:

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py
Comment (aboo-odoo) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:251]: @jco-odoo Isn't this breaking the stable policy ?

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py
Comment (aboo-odoo) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:251]: @jco-odoo Isn't this breaking the stable policy ?

Comment (jco-odoo) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:251]: It is in the grey zone as it is optional.

Comment (jco-odoo) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:251]: I am more wondering whether taxes_vals makes a lot of sense here.  Also because you can get it again from the line data in principle.  @hsal-odoo

Comment (jco-odoo) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:251]: It actually makes sense as you might need the taxes values in a lot of cases.

File: addons/l10n_jo_edi/models/account_move.py
Comment (jco-odoo) [reply addons/l10n_jo_edi/models/account_move.py:117]: @aboo-odoo There is a reason why this noqa is put: namely that we should not do this.  It is better for things to fail so we see the tracebacks than to fail silently.  I would not add the except Exception here.

Comment (aboo-odoo) [reply addons/l10n_jo_edi/models/account_move.py:117]: Yep, true my bad. There is actually a nicer way, `HTTPError` will never trigger since `RequestError` is the top level of `request` error implementation. So something like:
```suggestion
        try:
            response = requests.post(JOFOTARA_URL, data=params, headers=headers, timeout=50)
        except requests.exceptions.Timeout:
            return _("Request time out! Please try again.")
        except requests.exceptions.RequestException as e:
            return _("Invalid request: %s", e)
```

Would catch all requests errors other than a Timeout

**Comment by aboo-odoo** [comment]
@odoo/rd-security Hello security team, can you have a look on this ? The request in this PR uses a hardcoded URL and has a timeout set. The request must POST an XML document passed as parameter to the authority. The XML is generated just before the request which removes the risk of a user sending a malicious XML to the authority.

**Review by xmo-odoo — CHANGES_REQUESTED** [review]

File: addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py
Comment (hsal-odoo) [reply addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py:395]: How is the signature not respected?

**Comment by hsal-odoo** [comment]
@xmo-odoo can u kindly have a second round of review?

**Review by xmo-odoo — COMMENTED** [review]
@robodoo override=ci/security

**Comment by jco-odoo** [comment]
@Xavier-Do @d-fence Did you decide that new localization modules in stable also need approval?  (we might need a lot of approvals)  Do we just need your override here?  (see ci/template)

**Comment by Xavier-Do** [comment]
We didn't decide anything, we removed l10n from the blacklisted modules for upgrades so that localisation upgrades are tested as every other modules as requested by upgrade team.

As a side effect, exception will be needed to introduce new l10n modules yes, we may do something in the future to make it easier.

**Comment by Xavier-Do** [comment]
Upgrade exception [#406](https://runbot.odoo.com/web/#id=406&view_type=form&model=runbot.upgrade.exception) added.

Waiting the forward-port of this pr, the exception will be applied on all builds, and create an inconsistent state for migrations.
**Please forward port this asap up to master without change.** Exception should be forward-ported before the end of the week.

PRs adding fields/models should be merged at the beginning of the week and all the forward ports should be merged in the same week.

If you need to apply any change before it reaches master, please notify runbot team.

Details:
```
module:l10n_jo_edi
```
cc @aj-fuentes

**Comment by jco-odoo** [comment]
@robodoo r+

**Comment by jco-odoo** [comment]
@robodoo r+

File: addons/l10n_jo_edi/tests/jo_edi_common.py
Comment (jjmaksoud) [inline addons/l10n_jo_edi/tests/jo_edi_common.py:3]: @hsal-odoo `account_reports` is not a dependency, it's an enterprise module, can you update the test and make sure it runs with module dependencies only?

Comment (jco-odoo) [reply addons/l10n_jo_edi/tests/jo_edi_common.py:3]: Apparently it can just be replaced by AccountTestInvoicingCommon.setup_country

**Pull request closed without merging**
